### PR TITLE
feat: add new sector iterator function

### PR DESCRIPTION
**Pull request by shihang-zhang**
feat: add new sector iterator function to start iterating from last gc sector.)

### CODE REVIEW

File: src/fdb_kvdb.c [hunk 1]
```diff
@@ -1660,6 +1672,8 @@ fdb_err_t fdb_kvdb_init(fdb_kvdb_t db, const char *name, const char *path, struc
         db->default_kvs.num = 0;
         db->default_kvs.kvs = NULL;
     }
+
+    db->oldest_addr = 0;
```
Comment (armink) [inline src/fdb_kvdb.c:1720]: 第一次上电时要遍历整个数据区扇区，从而找到最旧的扇区，不能写死的

Comment (shihang-zhang) [reply src/fdb_kvdb.c:1720]: 请问应该怎样找到最旧的扇区？没有找到和这个信息相关的代码

Comment (armink) [reply src/fdb_kvdb.c:1720]: 这块有和 @weiguang-fan 沟通过，要不你们先同步一下吧

Comment (shihang-zhang) [reply src/fdb_kvdb.c:1720]: Hi armink, 我根据和weiguang讨论后的理解，提交了代码。请你帮忙review代码看看是否我对设计的实现有偏差。代码目前还未经过测试，只做讨论使用。Thanks！

**Comment by shihang-zhang** [comment]
Hi armink,
I have done a test and please refer to the following for some of the test log. The code runs as designed. 
Thanks!

[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 3, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 2, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00000000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00000800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00000800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00001000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00001000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00001800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00001800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00000000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00000000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00000800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00000800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00001000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00001000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00001800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00001800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00000000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00000000
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00000800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1425) Trigger a GC check after created KV.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1295) The remain empty sector is 1, GC threshold is 1.
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1265) Collect a sector @0x00000800
[FlashDB][kv][knx_os] (E:\3\0302lmg\Component\flashdb\src\FlashDB\src\fdb_kvdb.c:1277) oldest_addr is @0x00001000

File: src/fdb_kvdb.c [hunk 2]
```diff
@@ -1042,6 +1068,18 @@ static bool do_gc(kv_sec_info_t sector, void *arg1, void *arg2)
...
+        sec_addr = get_next_sector_addr(db, sector);
+        /*sec_addr reached db_max_size(db), roll back to the first sector*/
+        if (sec_addr == FAILED_ADDR)
+        {
+            db->oldest_addr = 0;
+        }
+        else
+        {
+            db->oldest_addr = sec_addr;
+        }
```
Comment (armink) [inline src/fdb_kvdb.c:1068]: L1074-L1081 的逻辑可以放到 get_next_sector_addr 里吗？

get_next_sector_addr 中检测到，下一个扇区如果超过最旧的扇区，则返回 FAILED_ADDR

这样 L821-L824 代码可能也就无需再做调整了

File: src/fdb_kvdb.c [hunk 3]
```diff
@@ -1450,6 +1488,44 @@ static void kv_auto_update(fdb_kvdb_t db)
...
+        // found the first using sector
+        if (arg->is_first_using_sector_found == false)
+        {
+            arg->is_first_using_sector_found = true;
+            arg->sector_using_addr = sector->addr;
+        }
+    }
+    else
+    {
+    }
```
Comment (armink) [inline src/fdb_kvdb.c]: 这里的 else 没用了可以删除

File: src/fdb_kvdb.c [hunk 4]
```diff
@@ -1450,6 +1488,44 @@ static void kv_auto_update(fdb_kvdb_t db)
...
+
+    return false;
+}
+
```
Comment (armink) [inline src/fdb_kvdb.c:1506]: 整体的代码风格希望和 FlashDB 原有的风格保持一致，否则会有很多无关改动。多种代码风格揉在一起，整体看起来也不太协调

**Review by armink — CHANGES_REQUESTED** [review]

File: src/fdb_kvdb.c (hunk 3 above)
Comment (shihang-zhang) [reply src/fdb_kvdb.c]: Done

File: src/fdb_kvdb.c (hunk 4 above)
Comment (shihang-zhang) [reply src/fdb_kvdb.c:1506]: 我对代码风格手动修改了一下，尽量和原风格保持一致。帮忙看看还有遗漏的地方吗？起因之前在我自己的vs code工程中修改过代码，导致部分代码被自动format了。多谢！

File: src/fdb_kvdb.c (hunk 2 above)
Comment (shihang-zhang) [reply src/fdb_kvdb.c:1068]: get_next_sector_addr在整个文件里面被多次调用，如果改动它的逻辑，也需要将调用它时候的起始sec_addr = 0;也做修改。尤其在fdb_kv_iterate函数里面iterate起始sector地址不是0，逻辑上没有想明白如果修改get_next_sector_addr后是否有其他影响？

Comment (armink) [reply src/fdb_kvdb.c:1068]: 好的，这块后面我来改改吧

**Comment by armink** [comment]
感谢 @shihang-zhang ，PR 已经合并了

**Pull request merged**